PerimeterXModule/Internals/CredentialsIntelligence/OutputFilterStream.cs

```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace PerimeterX
+{
+    /// <summary>
+    /// A stream which keeps an in-memory copy as it passes the bytes through
+    /// </summary>
+    public class OutputFilterStream : Stream
+    {
+        private readonly Stream InnerStream;
+        private readonly MemoryStream CopyStream;
+
+        public OutputFilterStream(Stream inner)
+        {
+            this.InnerStream = inner;
+            this.CopyStream = new MemoryStream();
+        }
+
+        public string ReadStream()
+        {
+            lock (this.InnerStream)
+            {
+                if (this.CopyStream.Length <= 0L ||
+                    !this.CopyStream.CanRead ||
+                    !this.CopyStream.CanSeek)
+                {
+                    return String.Empty;
+                }
+
+                long pos = this.CopyStream.Position;
+                this.CopyStream.Position = 0L;
+                try
+                {
+                    return new StreamReader(this.CopyStream).ReadToEnd();
+                }
+                finally
+                {
+                    try
+                    {
+                        this.CopyStream.Position = pos;
+                    }
+                    catch { }
+                }
+            }
+        }
+
+
+        public override bool CanRead
+        {
+            get { return this.InnerStream.CanRead; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return this.InnerStream.CanSeek; }
+        }
+
+        public override bool CanWrite
+        {
+            get { return this.InnerStream.CanWrite; }
+        }
+
+        public override void Flush()
+        {
+            this.InnerStream.Flush();
+        }
+
+        public override long Length
+        {
+            get { return this.InnerStream.Length; }
+        }
+
+        public override long Position
+        {
+            get { return this.InnerStream.Position; }
+            set { this.CopyStream.Position = this.InnerStream.Position = value; }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return this.InnerStream.Read(buffer, offset, count);
```

Comment:
Is this okay that it reads from the inner stream and does not update the copy stream?